File: Python/forest.py
from mushrooms import MushroomContext, MushroomFactory


class Forest:
    def __init__(self):
        self._mushrooms = []

    def get_mushrooms(self):
        return self._mushrooms

    def clear_mushrooms(self):
        self._mushrooms = []

    def add_mushroom(self, mushroom):
        self._mushrooms.append(mushroom)


class ForestRepository:
    def __init__(self, filename):
        self._filename = filename

    def add_mushroom_to_forest(self, name, eatable, treatment, forest: Forest):
        with open(self._filename, 'a') as file:
            file.write(f"\n{name} {eatable} {treatment}")
        self.load_mushrooms_list(forest)

    def load_mushrooms_list(self, forest):
        forest.clear_mushrooms()
        with open(self._filename, 'r') as file:
            for line in file:
                line = line.split(" ")
                context = MushroomContext(line[0], int(line[1]), int(line[2]))
                mushroom = MushroomFactory.create_mushroom(context)
                forest.add_mushroom(mushroom)

File: C++/OOP Lab4/forest.cpp
#include "forest.h"
#include <fstream>

Forest::~Forest() {
	for (auto mushroom : mushrooms) {
		delete mushroom;
	}
}

std::vector<Mushroom*> Forest::getMushrooms() {
	return mushrooms;
}
void Forest::addMushroom(Mushroom*& mushroom) {
	this->mushrooms.push_back(mushroom);
}
void Forest::clearMushrooms()
{
	this->mushrooms.clear();
}

ForestRepository::ForestRepository(std::string fname) : filename(fname) {};
void ForestRepository::addMushroomToForest(std::string name, bool eatable, int treatment, Forest& forest) {
	std::fstream file(filename, std::ios::app);
	if (!file.is_open()) {
		std::cerr << "Can't open file: " << filename << std::endl;
		system("pause");
		exit(0);
	}
	file << name << " " << eatable << " " << treatment << "\n";
	file.close();
	loadMushroomsList(forest);
}
void ForestRepository::loadMushroomsList(Forest& forest) {
	std::ifstream file(filename);
	if (!file.is_open()) {
		std::cerr << "Can't open file: " << filename << std::endl;
		system("pause");
		exit(0);
	}
	std::string name;
	bool eatable;
	int treatment;

	while (file >> name >> eatable >> treatment) {
		MushroomContext context(name, eatable, treatment);
		Mushroom* mushroom = MushroomFactory::createMushroom(context);
		forest.addMushroom(mushroom);
	}

	file.close();
}